### PR TITLE
feat(multiple api support): switching apis based on dataplane's supported version

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**
- Allow switching of api versions based on versions registered by dataplane
- Default to V0 api calls for older versions of dataplane
- Use latest api version of dataplane supports more than one api version.
- Make api version to be runtime as states, i,e can be updated at any point.

Signed-off-by: Abhinandan Purkait <abhinandan.purkait@mayadata.io>

### CODE REVIEW

File: control-plane/agents/core/src/core/grpc.rs [hunk 1]
```diff
@@ -79,38 +94,272 @@ impl GrpcContext {
...
+                Ok(Self {
+                    context: context.clone(),
+                    io_engine_v0: None,
+                    io_engine_v1: Some(MayaClientV1 { host }),
+                })
+            }
+        }
+    }
+
+    pub(crate) fn client_v0(&self) -> Result<MayaClientV0, SvcError> {
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/grpc.rs:172]: please doc comment these pub(crate).

File: control-plane/agents/core/src/core/grpc.rs [hunk 2]
```diff
@@ -154,6 +404,9 @@ impl GrpcClientLocked {
             client,
         })
     }
+    pub(crate) fn api_version(&self) -> APIVersion {
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/grpc.rs:418]: please doc comment

File: control-plane/agents/core/src/core/wrapper.rs [hunk 3]
```diff
@@ -105,38 +129,53 @@ impl NodeWrapper {
...
         timeout: NodeCommsTimeout,
     ) -> Result<GrpcContext, SvcError> {
-        GrpcContext::new(
-            self.lock.clone(),
-            self.id(),
-            &self.endpoint_str(),
-            &timeout,
-            None,
-        )
+        if self.latest_api_version().is_some() {
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/wrapper.rs]: if let Some(api_version) = self.latest_api_version() { ... }

File: control-plane/agents/core/src/core/wrapper.rs [hunk 4]
```diff
@@ -192,23 +231,40 @@ impl NodeWrapper {
...
+            self.comms_timeouts.connect(),
+            true,
+        );
+
+        // Set the api version to latest and make a call
+        self.node_state.api_versions = Some(vec![APIVersion::V1]);
+        let client = self.grpc_client_timeout(timeouts.clone()).await?;
+        match client.liveness_probe(self.id()).await {
+            Ok(data) => return Ok(data),
+            Err(_) => debug!("V1 liveness failed on startup, retrying with V0 liveness"),
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/wrapper.rs]: hmm wondering this should be an info?
nevertheless, could we improve the log by including the node id? eg:
```suggestion
            Err(_) => debug!(node.id=node_id, "V1 liveness failed on startup, retrying with V0 liveness"),
```

File: deployer/src/lib.rs [hunk 5]
```diff
@@ -273,6 +273,10 @@ pub struct StartOptions {
     /// Maximum number of concurrent rebuilds across the cluster.
     #[structopt(long)]
     max_rebuilds: Option<u32>,
+
+    /// api versions to be passed to the io-engine
+    #[structopt(long, env = "API_VERSION", default_value = "V0")]
+    api_versions: String,
```
Comment (tiagolobocastro) [inline deployer/src/lib.rs]: io_engine_api_versions? And env IO_ENGINE_API_VERSION

**Review by tiagolobocastro — APPROVED** [review]

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by tiagolobocastro** [comment]
bors cancel

**Comment by tiagolobocastro** [comment]
bors try

**Comment by tiagolobocastro** [comment]
bors cancel

**Comment by tiagolobocastro** [comment]
bors try

**Review by vharsh — APPROVED** [review]
Lgtm

**Review by Ab-hishek — APPROVED** [review]

**Comment by Abhinandan-Purkait** [comment]
bors try

File: control-plane/agents/core/src/core/wrapper.rs [hunk 6]
```diff
@@ -103,6 +103,18 @@ impl NodeWrapper {
...
     }
 
+    /// set the node state on apiversion change to the passed argument
+    pub(crate) fn set_state_on_version_change(&mut self, node_state: NodeState) {
+        if self.node_state().api_versions != node_state.api_versions {
+            debug!(
+                "API Version changed for node: {}, to : {:?}",
+                node_state.id.clone(),
+                node_state.api_versions
+            );
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/wrapper.rs]:
```rust
            debug!(
               node.id=%node_state.id,
                "API Versions changed from {:?} to {:?}",
                self.node_state().api_versions,
                node_state.api_versions
            );
```

Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/core/wrapper.rs]: Done.

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**